capreolus/trainer/__init__.py

```diff
@@ -225,20 +228,22 @@ def train(self, reranker, train_dataset, train_output_path, dev_data, dev_output
             # write model weights to file
             weights_fn = weights_output_path / f"{niter}.p"
             reranker.save_weights(weights_fn, self.optimizer)
-
+            iters_so_far += 1
             # predict performance on dev set
-            pred_fn = dev_output_path / f"{niter}.run"
-            preds = self.predict(reranker, dev_data, pred_fn)
 
-            # log dev metrics
-            metrics = evaluator.eval_runs(preds, qrels, ["ndcg_cut_20", "map", "P_20"])
-            logger.info("dev metrics: %s", " ".join([f"{metric}={v:0.3f}" for metric, v in sorted(metrics.items())]))
+            if iters_so_far % validation_frequency == 0:
```

Comment:
Maybe this should be based on `niters` rather than `iters_so_far`, so that the same iters are checked against the validation set even if `fastforward=True` causes training to be resumed?